ifrs17/Utils/IdentityPropertyReader.ipynb

```diff
@@ -0,0 +1,69 @@
+        {
+            "cell_type": "code",
+            "source": [
+                "#!import \"../DataModel/DataStructure\"",
+                "\nusing System.Linq.Expressions;",
+                "\nusing static Systemorph.Vertex.Equality.IdentityPropertyExtensions;"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "public class IdentityPropertyReader<T> where T : BaseVariableIdentity {",
+                "\n    private Func<T, string> identityReader;",
+                "\n    private string[] propertiesExcluded;",
+                "\n    ",
+                "\n    private IdentityPropertyReader(string[] props) {",
+                "\n        propertiesExcluded = props;",
+                "\n        identityReader = GetIdentityReader();",
+                "\n    }",
+                "\n",
+                "\n    private static Dictionary<string, IdentityPropertyReader<T>> InstancesByExcludedProperties = new();",
+                "\n",
+                "\n    private Func<T, string> GetIdentityReader() {",
+                "\n        var parameter = Expression.Parameter(typeof(T));",
+                "\n        var expression = typeof(T).GetIdentityProperties().Where(x => !propertiesExcluded.Contains(x.Name))",
+                "\n                                  .SelectMany(x => new Expression[]{ Expression.Constant(x.Name.ToString()+\":\"),",
+                "\n                                                   Expression.Call(typeof(IdentityPropertyReader<T>).GetMethod(nameof(GetString)), Expression.Convert(Expression.Property(parameter, x.Name), typeof(Object))) }",
+                "\n                                  ).Aggregate((x, y) => Expression.Call(typeof(IdentityPropertyReader<T>).GetMethod(nameof(IdentityPropertyReader<RawVariable>.Concat)), x, y) );",
+                "\n        return Expression.Lambda<Func<T,string>>(expression, parameter).Compile();",
+                "\n    }",
+                "\n",
+                "\n    public static IdentityPropertyReader<T> GetInstance(params string[] excludingProperties) {",
+                "\n        var key = string.Join(\",\", excludingProperties.OrderBy(x=>x));",
+                "\n        if(!InstancesByExcludedProperties.TryGetValue(key, out var instance))",
+                "\n            return InstancesByExcludedProperties[key] = new IdentityPropertyReader<T>(excludingProperties);",
+                "\n        return instance;",
+                "\n    }",
+                "\n    public string GetIdentityString(T x) => identityReader(x);",
+                "\n    ",
+                "\n    public static string Concat(string first, string second) => first + \" \" + second;",
+                "\n    public static string GetString(Object item) => (item == null) ? \"\" : item.ToString();",
+                "\n}"
```

Comment:
I don't like this method:
 1. this is very generic, and should not belong in a client project
 2. this has been implemented in other places before: platform V7, IFRS Generic V7 and it would be yet another version for the same functionality
 3. I think a few concepts must be discuss with the dev team, as they have already implemented similar concepts (see 2)
 4. I think this is not thread safe, with your static dictionary (line 41)
 5. Why do you reference RawVariable in a generic method (line 48)?
 6. In GetInstance, you seems to always return a new instance, meaning calling GetIdentityReader will always execute the reflection, I think.
 7. Converting all the string to then use in comparison/grouping is possible, but not so nice.

I summary, check this with the Dev team and they might already have a solution for this. If now, they can help you to write this better.

Reply:
Thanks! I proceeded with contacting the dev team.
Line 48 is a mistake, the method GetInstance works well. 
I did not find a better way of grouping than with a string. I wouldn't mind a suggestion :)

Reply:
The Dev team indeed gave a better way (ToString). See the new version.